src/components/bridge-blocker.ts

```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2021 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import * as logging from "./logging";
+const log = logging.get("BridgeBlocker");
+
+export class BridgeBlocker {
```

Comment:
Actually, what's the expected usage pattern here? Do bridges use this class directly?

Reply:
They might.

[IRC uses it as-is](https://github.com/matrix-org/matrix-appservice-irc/pull/1472/files#diff-2d366dcc8276d678ca71640bbe4bc4bc72320e4f54ac8c8f7f24642b1e2bf9f8R227)

[Slack extends it](https://github.com/matrix-org/matrix-appservice-slack/pull/612/files#diff-c9502d7239187f80ae0b64364678495cc78fd394448c552600f931ff8cc6d1f3R69) to disable its connection handlers as a blocking mechanism.

Reply:
Another dumb question, does this spit out a integer to Prometheus yet telling us if the bridge is blocked?

Reply:
Oops. No. Will fix that :)

Reply:
...actually, we can't easily do that without introducing an (optional) dependency on BridgeBlocker – currently we merely provide it for implementations but we don't force them to use it, or even assume that they might. Adding a `Bridge` opts/controller attribute just for the metric seems silly, so it may be better to just leave that up to the implementations.

Reply:
Now done in [IRC](https://github.com/matrix-org/matrix-appservice-irc/pull/1472/commits/c6aaf4d87b1258760f0c6913da8bcedba0105e60) and [Slack](https://github.com/matrix-org/matrix-appservice-slack/pull/612/commits/3ebd90a423efaf09565bbb76d4db4a7ad233d3c4)